### PR TITLE
Async calls implementation

### PR DESCRIPTION
**Pull request by smikis**
Will solve https://github.com/Vonage/vonage-dotnet-sdk/issues/126

### CODE REVIEW

**Review by slorello89 — CHANGES_REQUESTED** [review]
Hi @smikis - Thanks for this PR - it's a major undertaking. The issue is, as this is currently constituted it would be a breaking change to the SDK as folks would have to switch to the async implementation when upgrading. If this was an additive rather than a breaking change it would be something I'd like to use.

**Comment by smikis** [comment]
Any ideas how do you want to approach this? Just duplicating code for every method does not seem right to me

**Comment by smikis** [comment]
I'll try to wrap non async calls in Task.Run
It should work out fine

**Comment by smikis** [comment]
Update to include both methods. Async method is primary one, and non async method just calls async method and returns result.
Should I revert documentation and refer to non async methods?

**Review by slorello89 — CHANGES_REQUESTED** [review]
The way you implemented this I don't think we need to have the async methods called from the unit tests - should be sufficient for coverage to use the sync methods only so I'd ask that you revert the unit test updates (ignore code cov somewhat as it doesn't handle the branches created by the compiler for async that well)

File: Vonage.Test.Unit/AccountTest.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ public class AccountTest : TestBase
         [Theory]
         [InlineData(false)]
         [InlineData(true)]
-        public void GetAccountBalance(bool passCreds)
+        public async Task GetAccountBalance(bool passCreds)
```
[the remark this answers is not in the document]
Comment (smikis) [reply Vonage.Test.Unit/AccountTest.cs]: Sure, make sense. Reverted test files

File: Vonage/Request/ApiRequest.cs [hunk 2]
```diff
@@ -341,11 +342,11 @@ public static VonageResponse DoRequestWithUrlContent(string method, Uri uri, Dic
...
+        public static async Task<VonageResponse> SendHttpRequestAsync(HttpRequestMessage req)
         {
             var logger = Logger.LogProvider.GetLogger(LOGGER_CATEGORY);
-            var response = Configuration.Instance.Client.SendAsync(req).Result;
-            var stream = response.Content.ReadAsStreamAsync().Result;
+            var response = await Configuration.Instance.Client.SendAsync(req);
+            var stream = await response.Content.ReadAsStreamAsync();
             string json;
             using (var sr = new StreamReader(stream))
             {
```
Comment (slorello89) [inline Vonage/Request/ApiRequest.cs:352]: just below here it's reading out of the stream synchronously - that can be done using the Async/Await pattern as well

**Review by slorello89 — CHANGES_REQUESTED** [review]
Hi @smikis - I only see one more thing I'd ask you to change - other than that I'm going to do some manual tests on my end to validate everything still works outside of the unit tests - looking forward to having this in the SDK :)

File: Vonage/Request/ApiRequest.cs (hunk 2 above)
Comment (smikis) [reply Vonage/Request/ApiRequest.cs:352]: Updated

**Review by slorello89 — APPROVED** [review]
@smikis - Thank you so much for this - a stellar contribution. Happy Hacktoberfest! 🎃🎃🎃

**Pull request merged**